# Run with --help flag for help.
# Modified 08/01/2020 by Fabio Marroni
#We compared atlantica and marmorata transcripts. So the transcript "atlantica private" are those absent in marmorata.

suppressPackageStartupMessages({
  library(optparse)
})


option_list = list(
  make_option(c("-I", "--infile"), type="character", default="",
              help="File generated by SNPrelate IBD estimation to be used as input file", metavar="character"),
  make_option(c("-K", "--kinship"), type="numeric", default=0.05, 
              help="kinship to be considered significant [default= %default]", metavar="character"),
  make_option(c("-C", "--convfile"), type="character", default="", 
              help="File for converting names of different genome releases [default= %default]", metavar="character"),
  make_option(c("-G", "--gffile"), type="character", default="", 
              help="GFF file  [default= %default]", metavar="character"),
  make_option(c("-O", "--outfile"), type="character", default="", 
              help="output file name [default= %default]", metavar="character")
)

opt_parser = OptionParser(option_list=option_list);
opt = parse_args(opt_parser);

#print("USAGE: $ 01_small_RNA.r -I input dir/ -O summarized output file")

  if (is.null(opt$infile)) {
  stop("WARNING: No infile specified with '-I' flag.")
} else {  cat ("infile is ", opt$infile, "\n")
  infile <- opt$infile  
  }

  if (is.null(opt$kinship)) {
  stop("WARNING: No kinship specified with '-K' flag.")
} else {  cat ("kinship is ", opt$kinship, "\n")
  kinship <- opt$kinship  
  }

  if (is.null(opt$convfile)) {
  stop("WARNING: No convfile specified with '-C' flag.")
} else {  cat ("convfile is ", opt$convfile, "\n")
  convfile <- opt$convfile  
  }

  if (is.null(opt$gffile)) {
  stop("WARNING: No gffile specified with '-G' flag.")
} else {  cat ("gffile is ", opt$gffile, "\n")
  gffile <- opt$gffile  
  }

  if (is.null(opt$outfile)) {
  stop("WARNING: No outfile specified with '-O' flag.")
} else {  cat ("outfile is ", opt$outfile, "\n")
  outfile <- opt$outfile  
  }

  #The default value of kinship 0.05 was chosen because the 3rd quartile of kinship is 0.04, so we just stick slightly above it to be not too 
  #inclusive and not too conservative.
extract_wibd_sig<-function(infile,threshold,outfile,convfile,gffile,kinship)
{
library("data.table")
wibd<-fread(infile)
wibd$chrpos<-paste(wibd$chr,wibd$start,wibd$end,sep="_")
#Very nice way to aggregate and computing several summary statistics in one shot.
waggr<-wibd[,.(kinsum=sum(kinship),kinmed=median(kinship),kinmean=mean(kinship)),by=.(chrpos)]
waggr$chr<-unlist(lapply(strsplit(waggr$chrpos,"_"),"[",1))
waggr$start<-as.numeric(as.character(unlist(lapply(strsplit(waggr$chrpos,"_"),"[",2))))
waggr$end<-as.numeric(as.character(unlist(lapply(strsplit(waggr$chrpos,"_"),"[",3))))
waggr$window<-NA
waggr$kinmean[is.na(waggr$kinmean)]<-0
waggr$kinsum[is.na(waggr$kinsum)]<-0
waggr$kinmed[is.na(waggr$kinmed)]<-0
mywin<-0
#Stupid loop to only select windows in which the mean kinship is greater than our threshold
for(aaa in 2:nrow(waggr))
{
	if(waggr$chr[aaa-1]!=waggr$chr[aaa]) 
	{
		mywin<-mywin+1
		if(waggr$kinmean[aaa]>=kinship)	waggr$window[aaa]<-mywin
		next
	}
	if(waggr$kinmean[aaa]<kinship)
	{
		mywin<-mywin+1
		next
	}
	waggr$window[aaa]<-mywin
}
waggr<-waggr[!is.na(waggr$window),]
waggr2<-waggr[,.(chr=unique(chr),nwin=.N,start=min(start),end=max(end),mean.kinship=mean(kinmean)),by=.(window)]
waggr2<-waggr2[waggr2$nwin>1,]

#Merge the RefSeq chromosome names (because those are present in gff)
myconv<-fread(convfile,data.table=F)
myconv<-myconv[,c("RefSeq","INSDC")]
mysummary<-merge(waggr2,myconv,by.x="chr",by.y="INSDC",all.x=T,all.y=F,sort=F)
#Read the gff using scan, because the irregular format could create problems to fread.
#Why couldn't they create a GFF with a fixed number of columns? I don't know
mygff<-scan(gffile,what="",sep="\n")
#Skip header
mygff<-mygff[substr(mygff,1,1)!="#"]
#Split by tab and only get five fixed columns
ciccio<-strsplit(mygff,"\t")
mygff<-cbind(unlist(lapply(ciccio,"[",1)),unlist(lapply(ciccio,"[",3)),unlist(lapply(ciccio,"[",4)),unlist(lapply(ciccio,"[",5)),unlist(lapply(ciccio,"[",9)))
#I only keep the genes, we dont' care about exons
newgff<-mygff[mygff[,2]=="mRNA",]
newgff<-data.frame(newgff,stringsAsFactors=F)
newgff$X2<-NULL
#Extract gene names and gene ID from the gff
newgff$genbank<-unlist(lapply(strsplit(newgff$X5,";"),"[",3))
newgff$genbank<-gsub("Genbank:","",unlist(lapply(strsplit(newgff$genbank,","),"[",2)))
#newgff$gene_name<-gsub("gene=","",unlist(lapply(strsplit(newgff$X5,";"),"[",6)))
newgff$gene_name<-unlist(lapply(strsplit(newgff$X5,"Dbxref=GeneID:"),"[",2))
newgff$gene_name<-unlist(lapply(strsplit(newgff$gene_name,","),"[",1))
newgff$description<-unlist(lapply(strsplit(newgff$X5,"product="),"[",2))
newgff$description<-unlist(lapply(strsplit(newgff$description,";"),"[",1))
newgff$description<-unlist(lapply(strsplit(newgff$description,"%"),"[",1))
#newgff$gene_ID<-gsub("Dbxref=GeneID:","",unlist(lapply(strsplit(newgff$X5,";"),"[",2)))
newgff$X5<-NULL
setnames(newgff,c("X1","X3","X4"),c("chr","start","end"))
newgff$start<-as.numeric(as.character(newgff$start))
newgff$end<-as.numeric(as.character(newgff$end))
mysummary$description<-mysummary$gene_name<-mysummary$genbank<-""
for(bbb in 1:nrow(mysummary))
{
	cat("Riga",bbb,"di",nrow(mysummary),"\n")
	sgff<-newgff[newgff$chr%in%mysummary$RefSeq[bbb],]
	sgff$ov<-NA
	#Identify genes overlapping the region with significant HFLK signal
	for(ccc in 1:nrow(sgff))
	{
	sgff$ov[ccc]<-count.overlap(sgff$start[ccc],sgff$end[ccc],mysummary$start[bbb],mysummary$end[bbb])
	}
	#Print all the genes names located in each significant window
	mysummary$genbank[bbb]<-paste(unique(sgff$genbank[sgff$ov>0]),sep=";",collapse=";")
	mysummary$gene_name[bbb]<-paste(unique(sgff$gene_name[sgff$ov>0]),sep=";",collapse=";")
	mysummary$description[bbb]<-paste(unique(sgff$description[sgff$ov>0]),sep=";",collapse=";")
}

write.table(mysummary,outfile,quote=F,sep="\t",row.names=F)
}




count.overlap<-function(start.1=1,end.1=4845859,start.2=1,end.2=84945)
{
	max.start<-max(start.1,start.2)
	min.end<-min(end.1,end.2)
	overlap<-max(0,(1+min.end-max.start))
	overlap
}

count.overlap.vector<-function(x)
{
	max.start<-max(x[1],x[3])
	min.end<-min(x[2],x[4])
	overlap<-max(0,(1+min.end-max.start))
	overlap
}


extract_wibd_sig(infile=infile,kinship=kinship,outfile=outfile,convfile=convfile,gffile=gffile)
